Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/D3D12Quad.cpp

```diff
@@ -0,0 +1,832 @@
+    {
+        // if image format needed to be converted, the wic converter will contain the converted image
+        hr = wicConverter->CopyPixels(0, bytesPerRow, imageSize, *imageData);
+        if (FAILED(hr)) return 0;
+    }
+    else
+    {
+        // no need to convert, just copy data from the wic frame
+        hr = wicFrame->CopyPixels(0, bytesPerRow, imageSize, *imageData);
+        if (FAILED(hr)) return 0;
+    }
+
+    // now describe the texture with the information we have obtained from the image
+    resourceDescription = {};
+    resourceDescription.Dimension = D3D12_RESOURCE_DIMENSION_TEXTURE2D;
+    resourceDescription.Alignment = 0; // may be 0, 4KB, 64KB, or 4MB. 0 will let runtime decide between 64KB and 4MB (4MB for multi-sampled textures)
+    resourceDescription.Width = textureWidth; // width of the texture
+    resourceDescription.Height = textureHeight; // height of the texture
+    resourceDescription.DepthOrArraySize = 1; // if 3d image, depth of 3d image. Otherwise an array of 1D or 2D textures (we only have one image, so we set 1)
+    resourceDescription.MipLevels = 1; // Number of mipmaps. We are not generating mipmaps for this texture, so we have only one level
+    resourceDescription.Format = dxgiFormat; // This is the dxgi format of the image (format of the pixels)
+    resourceDescription.SampleDesc.Count = 1; // This is the number of samples per pixel, we just want 1 sample
+    resourceDescription.SampleDesc.Quality = 0; // The quality level of the samples. Higher is better quality, but worse performance
+    resourceDescription.Layout = D3D12_TEXTURE_LAYOUT_UNKNOWN; // The arrangement of the pixels. Setting to unknown lets the driver choose the most efficient one
+    resourceDescription.Flags = D3D12_RESOURCE_FLAG_NONE; // no flags
+
+    // return the size of the image.
+    return imageSize;
+}
+
+void D3D12Quad::ThrowIfFailed(HRESULT hr)
+{
+    if (FAILED(hr))
+    {
+        throw HrException(hr);
+    }
+}
+
+void D3D12Quad::GetHardwareAdapter(
+    IDXGIFactory1* pFactory,
+    IDXGIAdapter1** ppAdapter,
+    bool requestHighPerformanceAdapter)
+{
+    *ppAdapter = nullptr;
+
+    ComPtr<IDXGIAdapter1> adapter;
+
+    ComPtr<IDXGIFactory6> factory6;
+    if (SUCCEEDED(pFactory->QueryInterface(IID_PPV_ARGS(&factory6))))
+    {
+        for (
+            UINT adapterIndex = 0;
+            SUCCEEDED(factory6->EnumAdapterByGpuPreference(
+                adapterIndex,
+                requestHighPerformanceAdapter == true ? DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE : DXGI_GPU_PREFERENCE_UNSPECIFIED,
+                IID_PPV_ARGS(&adapter)));
+            ++adapterIndex)
+        {
+            DXGI_ADAPTER_DESC1 desc;
+            adapter->GetDesc1(&desc);
+
+            if (desc.Flags & DXGI_ADAPTER_FLAG_SOFTWARE)
+            {
+                // Don't select the Basic Render Driver adapter.
+                // If you want a software adapter, pass in "/warp" on the command line.
```

Comment:
>pass in "/warp" on the command line. [](http://example.com/codeflow?start=51&length=36)

Does this app accept command line parameters? I didn't see any in the main entry point.

Reply:
Nope. Removing comment.